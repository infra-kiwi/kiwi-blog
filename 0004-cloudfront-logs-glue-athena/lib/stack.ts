import * as cdk from 'aws-cdk-lib';
import { aws_cloudfront, aws_s3, CfnOutput, RemovalPolicy, Stack } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { HttpOrigin } from 'aws-cdk-lib/aws-cloudfront-origins';
import * as aws_glue_alpha from '@aws-cdk/aws-glue-alpha';

export class CloudFrontStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Create a dedicated S3 bucket to store CloudFront logs
    const logBucket = new aws_s3.Bucket(this, 'CloudFrontLogs', {
      // ACL, required by CloudFront
      objectOwnership: aws_s3.ObjectOwnership.BUCKET_OWNER_PREFERRED,
      // Make sure we don't keep unnecessary data after we delete this stack
      removalPolicy: RemovalPolicy.DESTROY,
      autoDeleteObjects: true
    });

    const distribution = new aws_cloudfront.Distribution(this, 'Distribution', {
      // HTTPS (because we should ALWAYS use HTTPS!)
      sslSupportMethod: aws_cloudfront.SSLMethod.SNI,

      // Enable logging to the S3 bucket
      enableLogging: true,
      logBucket: logBucket,

      // Use the wonderful https://reqres.in/ as a destination for our calls
      defaultBehavior: {
        origin: new HttpOrigin('reqres.in')
      }
    });
    // This output will contain the domain name of the distribution,
    // e.g. xxxxxxx.cloudfront.net
    new CfnOutput(this, 'DistributionDomainName', {
      value: distribution.domainName
    });

    // Create a database where to store the table
    const database = new aws_glue_alpha.Database(this, 'GlueDatabase', {
      // Give the database a friendly name (lower-case only!)
      databaseName: Stack.of(this).stackName + '-database'
    });
    new CfnOutput(this, 'CloudFrontLogsDatabaseName', {
      value: database.databaseName
    });

    // Create the table that maps each log line
    const table = new aws_glue_alpha.S3Table(this, 'CloudFrontLogsTable', {
      // Give the table a friendly name (lower-case only!)
      tableName: Stack.of(this).stackName + '-cloudfrontlogstable',

      // The source bucket for this Glue table
      bucket: logBucket,
      database: database,

      // Full SQL reference: https://docs.aws.amazon.com/athena/latest/ug/cloudfront-logs.html
      columns: [
        { name: 'date', type: aws_glue_alpha.Schema.DATE },
        { name: 'time', type: aws_glue_alpha.Schema.STRING },
        { name: 'x_edge_location', type: aws_glue_alpha.Schema.STRING },
        { name: 'sc_bytes', type: aws_glue_alpha.Schema.BIG_INT },
        { name: 'c_ip', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_method', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_host', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_uri_stem', type: aws_glue_alpha.Schema.STRING },
        { name: 'sc_status', type: aws_glue_alpha.Schema.INTEGER },
        { name: 'cs_referrer', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_user_agent', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_uri_query', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_cookie', type: aws_glue_alpha.Schema.STRING },
        { name: 'x_edge_result_type', type: aws_glue_alpha.Schema.STRING },
        { name: 'x_edge_request_id', type: aws_glue_alpha.Schema.STRING },
        { name: 'x_host_header', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_protocol', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_bytes', type: aws_glue_alpha.Schema.BIG_INT },
        { name: 'time_taken', type: aws_glue_alpha.Schema.FLOAT },
        { name: 'x_forwarded_for', type: aws_glue_alpha.Schema.STRING },
        { name: 'ssl_protocol', type: aws_glue_alpha.Schema.STRING },
        { name: 'ssl_cipher', type: aws_glue_alpha.Schema.STRING },
        { name: 'x_edge_response_result_type', type: aws_glue_alpha.Schema.STRING },
        { name: 'cs_protocol_version', type: aws_glue_alpha.Schema.STRING },
        { name: 'fle_status', type: aws_glue_alpha.Schema.STRING },
        { name: 'fle_encrypted_fields', type: aws_glue_alpha.Schema.INTEGER },
        { name: 'c_port', type: aws_glue_alpha.Schema.INTEGER },
        { name: 'time_to_first_byte', type: aws_glue_alpha.Schema.FLOAT },
        { name: 'x_edge_detailed_result_type', type: aws_glue_alpha.Schema.STRING },
        { name: 'sc_content_type', type: aws_glue_alpha.Schema.STRING },
        { name: 'sc_content_len', type: aws_glue_alpha.Schema.BIG_INT },
        { name: 'sc_range_start', type: aws_glue_alpha.Schema.BIG_INT },
        { name: 'sc_range_end', type: aws_glue_alpha.Schema.BIG_INT }
      ],

      // Define the right settings to parse CloudFront logs
      dataFormat: aws_glue_alpha.DataFormat.TSV,
      parameters: {
        // THESE PARAMETERS ARE VERY IMPORTANT
        'skip.header.line.count': '2',
        'serialization.format': '\t',
        'field.delim': '\t'
      }
    });
    new CfnOutput(this, 'CloudFrontLogsTableName', {
      value: table.tableName
    });
  }
}
