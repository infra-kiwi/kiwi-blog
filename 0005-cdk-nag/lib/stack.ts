/*
 * Copyright (c) 2024. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

import * as cdk from 'aws-cdk-lib';
import { aws_kms, aws_sns, CfnOutput } from 'aws-cdk-lib';
import { Construct } from 'constructs';

export class KiwiStack extends cdk.Stack {
  readonly topic: aws_sns.Topic;

  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Use the default AWS-provided SNS KMS key
    const awsSNSKey = aws_kms.Alias.fromAliasName(
      this,
      'AWSManagedSNSKey',
      'alias/aws/sns'
    );

    this.topic = new aws_sns.Topic(this, 'MyTopic', {
      masterKey: awsSNSKey
    });
    new CfnOutput(this, 'MyTopicArn', {
      value: this.topic.topicArn
    });
  }
}
