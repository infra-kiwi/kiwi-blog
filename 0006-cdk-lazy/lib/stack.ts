import * as cdk from 'aws-cdk-lib';
import { aws_sns, Stack } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { MultiLineParameter } from './multi-line-parameter';

export class KiwiStack extends cdk.Stack {
  readonly topic: aws_sns.Topic;

  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const myParameter = new MultiLineParameter(this, 'MyParameter', {
      parameterName: Stack.of(this).stackName + '-my-parameter'
    });

    // Populate the lines of our multi-line parameter
    myParameter.addLine("It's not DNS");
    myParameter.addLine("There's no way it's DNS");
    myParameter.addLine('It was DNS');

    // Add a dynamic line that depends on another CDK resource's value
    const aTopicAboutDNS = new aws_sns.Topic(this, 'ATopicAboutDNS');
    myParameter.addLine(
      `Here is a topic about DNS: ${aTopicAboutDNS.topicArn}`
    );
    this.topic = aTopicAboutDNS;
  }
}
