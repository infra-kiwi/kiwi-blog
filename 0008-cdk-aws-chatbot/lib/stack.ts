/*
 * Copyright (c) 2024. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { ChatBotSlackChannel } from './constructs/chatbot-slack-channel';
import { PostToSlack } from './constructs/post-to-slack';
import { Config } from './util/config';

export type ChatBotStackProps = cdk.StackProps & {
  config: Config;
};

export class ChatBotStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: ChatBotStackProps) {
    super(scope, id, props);

    const chatBotSlackChannels: Record<string, ChatBotSlackChannel> = {};

    // Create all the possible ChatBot channel configurations. To do so,
    // extract all available Slack workspace/channel pairs, deduplicate them,
    // and create a ChatBotSlackChannel for each
    const slackChannelEntries = Array.from(
      new Set(
        Object.entries(props.config.notificationChannels).map(([, v]) =>
          JSON.stringify([v.slackWorkspaceId, v.slackChannelId])
        )
      )
    ).map((e) => JSON.parse(e)) as [string, string][];

    for (const entry of slackChannelEntries) {
      const [slackWorkspaceId, slackChannelId] = entry;

      // Configure the channel
      const channel = new ChatBotSlackChannel(
        this,
        `ChatBotSlackChannel-${slackWorkspaceId}-${slackChannelId}`,
        {
          slackWorkspaceId,
          slackChannelId
        }
      );
      chatBotSlackChannels[JSON.stringify(entry)] = channel;
    }

    // Create all the PostToSlack instances, which will accept our custom payloads
    // and translate them into ChatBot-formatted ones
    for (const notificationChannelName in props.config.notificationChannels) {
      const config = props.config.notificationChannels[notificationChannelName];

      // Use the previously-generated Slack channel configuration
      const chatBotSlackChannel =
        chatBotSlackChannels[
          JSON.stringify([config.slackWorkspaceId, config.slackChannelId])
        ];

      new PostToSlack(this, `SlackLambda-${notificationChannelName}`, {
        chatbotSNSTopicArn: chatBotSlackChannel.topicArn,
        notificationChannelName
      });
    }
  }
}
