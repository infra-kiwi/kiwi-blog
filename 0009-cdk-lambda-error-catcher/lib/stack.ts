import * as cdk from 'aws-cdk-lib';
import { aws_lambda, CfnOutput, Stack } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import path from 'node:path';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { ErrorCatcherLambda } from './constructs/error-catcher-lambda';
import { LoggingFormat } from 'aws-cdk-lib/aws-lambda';

export type KiwiStackProps = cdk.StackProps & {
  // Expects an SNS topic that accepts a {title,body} payload, like the one configured in
  // https://blog.infra.kiwi/aws-cdk-a-slack-notification-pipeline-via-aws-chatbot-0e0c76e7f4c3
  errorCatcherSNSTopicArn?: string;
};

export class KiwiStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: KiwiStackProps) {
    super(scope, id, props);

    // Define our error catcher Lambda, which will listen
    // for error messages in logs
    const errorCatcherLambda = new ErrorCatcherLambda(this, 'Catch', {
      snsTopicArn: props.errorCatcherSNSTopicArn
    });

    // Define a test Lambda that will generate errors and log
    // them using the default TEXT format (props not specified)
    const lambdaWithTextLogs = new NodejsFunction(this, 'TestLambdaTextLogs', {
      functionName: Stack.of(this).stackName + 'test-lambda-text',
      entry: path.join(__dirname, 'lambda/test.ts'),
      runtime: aws_lambda.Runtime.NODEJS_20_X,
      bundling: {
        externalModules: ['@aws-sdk/*'],
        minify: false
      },
      // The Lambda entrypoint name
      handler: 'index.handler'
    });
    new CfnOutput(this, 'TestLambdaTextLogsFunctionName', {
      value: lambdaWithTextLogs.functionName
    });

    // Creates a subscription filter for the LogGroup of the lambdaWithTextLogs
    // function and forwards all errors to the errorCatcherLambda
    errorCatcherLambda.registerErrorCatcher(lambdaWithTextLogs);

    // Define a test Lambda that will generate errors and log
    // them using the JSON format
    const lambdaWithJSONLogs = new NodejsFunction(this, 'TestLambdaJSONLogs', {
      functionName: Stack.of(this).stackName + 'test-lambda-json',
      entry: path.join(__dirname, 'lambda/test.ts'),
      runtime: aws_lambda.Runtime.NODEJS_20_X,
      bundling: {
        externalModules: ['@aws-sdk/*'],
        minify: false
      },
      // The Lambda entrypoint name
      handler: 'index.handler',
      // Define the JSON format for logs
      loggingFormat: LoggingFormat.JSON
    });
    new CfnOutput(this, 'TestLambdaJSONLogsFunctionName', {
      value: lambdaWithJSONLogs.functionName
    });

    // Creates a subscription filter for the LogGroup of the lambdaWithJSONLogs
    // function and forwards all errors to the errorCatcherLambda
    errorCatcherLambda.registerErrorCatcher(lambdaWithJSONLogs);
  }
}
