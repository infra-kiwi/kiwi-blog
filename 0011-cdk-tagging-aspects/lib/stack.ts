/*
 * Copyright (c) 2024. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

import * as cdk from 'aws-cdk-lib';
import { aws_sns, CfnOutput } from 'aws-cdk-lib';
import { Construct } from 'constructs';

export class KiwiStack extends cdk.Stack {
  readonly topic: aws_sns.Topic;

  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    this.topic = new aws_sns.Topic(this, 'MyTopic');

    // Define an output for debugging purposes
    new CfnOutput(this, 'MyTopicArnOutput', {
      value: this.topic.topicArn
    });
  }
}
