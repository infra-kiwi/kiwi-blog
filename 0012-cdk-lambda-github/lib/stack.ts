/*
 * Copyright (c) 2024. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

import * as cdk from "aws-cdk-lib";
import {
  aws_apigateway,
  aws_lambda,
  aws_lambda_nodejs,
  aws_secretsmanager,
  CfnOutput,
  Duration,
  Stack,
} from "aws-cdk-lib";
import { Construct } from "constructs";
import path from "node:path";
import {
  secretNameGitHubAppId,
  secretNameGitHubAppInstallationId,
  secretNameGitHubPrivateKey,
} from "./lambda/common-secret";

export class KiwiStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Create a GitHub webhook token. This secret will be automatically
    // filled by Secrets Manager with a random string
    const webhookToken = new aws_secretsmanager.Secret(
      this,
      "GitHubWebhookToken",
      {
        secretName: `${Stack.of(this).stackName}-webhook-token`,
      },
    );
    new CfnOutput(this, "GitHubWebHookTokenSecretName", {
      value: webhookToken.secretName,
    });

    // Define the webhook handler lambda function
    const webhookLambda = new aws_lambda_nodejs.NodejsFunction(
      this,
      "GitHubWebHookLambda",
      {
        functionName: `${Stack.of(this).stackName}-webhook-lambda`,
        entry: path.join(__dirname, "lambda/pr-love.ts"),
        runtime: aws_lambda.Runtime.NODEJS_20_X,
        handler: "index.handler",
        bundling: {
          externalModules: ["@aws-sdk/*"],
        },
        environment: {
          GITHUB_WEBHOOK_TOKEN_SECRET_NAME: webhookToken.secretName,
        },
        // When we interact with the GitHub API, be prepared for longer
        // waiting times!
        timeout: Duration.seconds(10),
      },
    );

    // Grant access to all the various secrets
    webhookToken.grantRead(webhookLambda);
    [
      secretNameGitHubAppId,
      secretNameGitHubAppInstallationId,
      secretNameGitHubPrivateKey,
    ].forEach((secretName) =>
      aws_secretsmanager.Secret.fromSecretNameV2(
        this,
        `Secret-${secretName}`,
        secretName,
      ).grantRead(webhookLambda),
    );

    // Create a new API gateway, which will be the entrypoint for our webhooks
    const gateway = new aws_apigateway.RestApi(this, "GitHubWebHookGateway");

    // Create the webhook handler route /webhook
    const webhook = gateway.root
      .addResource("webhook")
      .addMethod("POST", new aws_apigateway.LambdaIntegration(webhookLambda));

    new CfnOutput(this, "GitHubWebHookUrl", {
      value: gateway.urlForPath(webhook.resource.path),
    });
  }
}
