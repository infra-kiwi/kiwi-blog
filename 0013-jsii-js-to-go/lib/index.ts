/*
 * Copyright (c) 2025. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

export class Constants {
  static readonly name = 'Mario';
}

export interface Home {
  readonly address: string;
}

export class Addressor {
  getFullGreeting(name: string, home: Home) {
    return `Hello ${Constants.name}! You live at ${home.address}.`;
  }
}
