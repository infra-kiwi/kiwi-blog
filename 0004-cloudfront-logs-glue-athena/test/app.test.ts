import { CloudFrontStack } from '../lib/stack';
import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';

test('KiwiStack', async () => {
  const app = new App();

  // Init stack
  const stack = new CloudFrontStack(app, 'Test');

  // Prepare for assertions
  const template = Template.fromStack(stack);
  expect(template).toMatchSnapshot();

  template.resourceCountIs('AWS::S3::Bucket', 1);
  template.resourceCountIs('AWS::CloudFront::Distribution', 1);
  template.resourceCountIs('AWS::Glue::Database', 1);
  template.resourceCountIs('AWS::Glue::Table', 1);
});
