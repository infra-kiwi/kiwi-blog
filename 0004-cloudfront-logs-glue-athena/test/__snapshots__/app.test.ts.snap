// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KiwiStack 1`] = `
{
  "Outputs": {
    "CloudFrontLogsDatabaseName": {
      "Value": {
        "Ref": "GlueDatabase24880C74",
      },
    },
    "CloudFrontLogsTableName": {
      "Value": {
        "Ref": "CloudFrontLogsTable7B7D3D61",
      },
    },
    "DistributionDomainName": {
      "Value": {
        "Fn::GetAtt": [
          "Distribution830FAC52",
          "DomainName",
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CloudFrontLogs13CB081E": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "OwnershipControls": {
          "Rules": [
            {
              "ObjectOwnership": "BucketOwnerPreferred",
            },
          ],
        },
        "Tags": [
          {
            "Key": "aws-cdk:auto-delete-objects",
            "Value": "true",
          },
        ],
      },
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Delete",
    },
    "CloudFrontLogsAutoDeleteObjectsCustomResource1A86C2D9": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "CloudFrontLogsPolicy7447C3A8",
      ],
      "Properties": {
        "BucketName": {
          "Ref": "CloudFrontLogs13CB081E",
        },
        "ServiceToken": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F",
            "Arn",
          ],
        },
      },
      "Type": "Custom::S3AutoDeleteObjects",
      "UpdateReplacePolicy": "Delete",
    },
    "CloudFrontLogsPolicy7447C3A8": {
      "Properties": {
        "Bucket": {
          "Ref": "CloudFrontLogs13CB081E",
        },
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:PutBucketPolicy",
                "s3:GetBucket*",
                "s3:List*",
                "s3:DeleteObject*",
              ],
              "Effect": "Allow",
              "Principal": {
                "AWS": {
                  "Fn::GetAtt": [
                    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
                    "Arn",
                  ],
                },
              },
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "CloudFrontLogs13CB081E",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "CloudFrontLogs13CB081E",
                          "Arn",
                        ],
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::S3::BucketPolicy",
    },
    "CloudFrontLogsTable7B7D3D61": {
      "Properties": {
        "CatalogId": {
          "Ref": "AWS::AccountId",
        },
        "DatabaseName": {
          "Ref": "GlueDatabase24880C74",
        },
        "TableInput": {
          "Description": "Test-cloudfrontlogstable generated by CDK",
          "Name": "Test-cloudfrontlogstable",
          "Parameters": {
            "field.delim": "	",
            "has_encrypted_data": true,
            "serialization.format": "	",
            "skip.header.line.count": "2",
          },
          "StorageDescriptor": {
            "Columns": [
              {
                "Name": "date",
                "Type": "date",
              },
              {
                "Name": "time",
                "Type": "string",
              },
              {
                "Name": "x_edge_location",
                "Type": "string",
              },
              {
                "Name": "sc_bytes",
                "Type": "bigint",
              },
              {
                "Name": "c_ip",
                "Type": "string",
              },
              {
                "Name": "cs_method",
                "Type": "string",
              },
              {
                "Name": "cs_host",
                "Type": "string",
              },
              {
                "Name": "cs_uri_stem",
                "Type": "string",
              },
              {
                "Name": "sc_status",
                "Type": "int",
              },
              {
                "Name": "cs_referrer",
                "Type": "string",
              },
              {
                "Name": "cs_user_agent",
                "Type": "string",
              },
              {
                "Name": "cs_uri_query",
                "Type": "string",
              },
              {
                "Name": "cs_cookie",
                "Type": "string",
              },
              {
                "Name": "x_edge_result_type",
                "Type": "string",
              },
              {
                "Name": "x_edge_request_id",
                "Type": "string",
              },
              {
                "Name": "x_host_header",
                "Type": "string",
              },
              {
                "Name": "cs_protocol",
                "Type": "string",
              },
              {
                "Name": "cs_bytes",
                "Type": "bigint",
              },
              {
                "Name": "time_taken",
                "Type": "float",
              },
              {
                "Name": "x_forwarded_for",
                "Type": "string",
              },
              {
                "Name": "ssl_protocol",
                "Type": "string",
              },
              {
                "Name": "ssl_cipher",
                "Type": "string",
              },
              {
                "Name": "x_edge_response_result_type",
                "Type": "string",
              },
              {
                "Name": "cs_protocol_version",
                "Type": "string",
              },
              {
                "Name": "fle_status",
                "Type": "string",
              },
              {
                "Name": "fle_encrypted_fields",
                "Type": "int",
              },
              {
                "Name": "c_port",
                "Type": "int",
              },
              {
                "Name": "time_to_first_byte",
                "Type": "float",
              },
              {
                "Name": "x_edge_detailed_result_type",
                "Type": "string",
              },
              {
                "Name": "sc_content_type",
                "Type": "string",
              },
              {
                "Name": "sc_content_len",
                "Type": "bigint",
              },
              {
                "Name": "sc_range_start",
                "Type": "bigint",
              },
              {
                "Name": "sc_range_end",
                "Type": "bigint",
              },
            ],
            "Compressed": false,
            "InputFormat": "org.apache.hadoop.mapred.TextInputFormat",
            "Location": {
              "Fn::Join": [
                "",
                [
                  "s3://",
                  {
                    "Ref": "CloudFrontLogs13CB081E",
                  },
                  "/",
                ],
              ],
            },
            "OutputFormat": "org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat",
            "SerdeInfo": {
              "SerializationLibrary": "org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe",
            },
            "StoredAsSubDirectories": false,
          },
          "TableType": "EXTERNAL_TABLE",
        },
      },
      "Type": "AWS::Glue::Table",
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderHandler9D90184F": {
      "DependsOn": [
        "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "6c1e9b465fa4b2d651dbc9ce3e732d8702a7b19137327684a71d89f1d355f1a2.zip",
        },
        "Description": {
          "Fn::Join": [
            "",
            [
              "Lambda function for auto-deleting objects in ",
              {
                "Ref": "CloudFrontLogs13CB081E",
              },
              " S3 bucket.",
            ],
          ],
        },
        "Handler": "index.handler",
        "MemorySize": 128,
        "Role": {
          "Fn::GetAtt": [
            "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "CustomS3AutoDeleteObjectsCustomResourceProviderRole3B1BD092": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "Distribution830FAC52": {
      "Properties": {
        "DistributionConfig": {
          "DefaultCacheBehavior": {
            "CachePolicyId": "658327ea-f89d-4fab-a63d-7e88639e58f6",
            "Compress": true,
            "TargetOriginId": "TestDistributionOrigin1552FC039",
            "ViewerProtocolPolicy": "allow-all",
          },
          "Enabled": true,
          "HttpVersion": "http2",
          "IPV6Enabled": true,
          "Logging": {
            "Bucket": {
              "Fn::GetAtt": [
                "CloudFrontLogs13CB081E",
                "RegionalDomainName",
              ],
            },
          },
          "Origins": [
            {
              "CustomOriginConfig": {
                "OriginProtocolPolicy": "https-only",
                "OriginSSLProtocols": [
                  "TLSv1.2",
                ],
              },
              "DomainName": "reqres.in",
              "Id": "TestDistributionOrigin1552FC039",
            },
          ],
        },
      },
      "Type": "AWS::CloudFront::Distribution",
    },
    "GlueDatabase24880C74": {
      "Properties": {
        "CatalogId": {
          "Ref": "AWS::AccountId",
        },
        "DatabaseInput": {
          "Name": "Test-database",
        },
      },
      "Type": "AWS::Glue::Database",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
